RUN-2993: autocomplete

jayas006 opened the pull request:
<!--
IMPORTANT: Before submitting your Pull Request, please review the following instructions:

1. Please follow the [Developer Guidelines](https://github.com/rundeck/rundeck/wiki/Developer-Guidelines) document.
2. Are you implementing a feature or enhancement? Please search the existing Issues and look 
   at the [Rundeck Roadmap Trello Board](https://trello.com/b/sn3g9nOr/rundeck-development) for your idea before posting.
   If your enhancement is not listed, it is better to 
   [post a new enhancement request](https://github.com/rundeck/rundeck/issues/new?template=feature_request.md)
   and get feedback from maintainers and the community *before* submitting an Pull request to implement it.
3. Please be sure the issue you are addressing is referenced in a commit, or the body of your PR,
   using [github keywords](https://help.github.com/articles/closing-issues-using-keywords/), e.g. "fixes #123".
-->

**Is this a bugfix, or an enhancement? Please describe.**
<!-- Include a clear and concise description of the bug you are fixing (reference issue number), or enhancement you are implementing. -->
enhancement
**Describe the solution you've implemented**
<!-- A clear and concise description of what you want to happen. -->

**Describe alternatives you've considered**
<!-- A clear and concise description of any alternative solutions or features you've considered. -->

**Additional context**
<!-- Add any other context or screenshots about the change here. -->

smartinellibenedetti commented:
@jayas006 could you please rebase this PR? I see that the rundeck-tokens.css is an outdated version

smartinellibenedetti commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/components/primeVue/PtAutoComplete/tests/PtAutoComplete.spec.ts:
```diff
@@ -0,0 +1,72 @@
+            ...props,
+        },
+        global: {
+            stubs: {
+                AutoComplete: {
+                    template: `
+                        <div>
+                            <input
+                                class='p-inputtext'
+                                @input='$emit("update:modelValue", $event.target.value)'
+                                @change='$emit("onChange", $event)'
+                                @complete='$emit("complete", "test suggestion")'
+                            />
+                        </div>
+                    `,
+                },
+            },
```
```suggestion
```

Once I pushed the fix for the jest.config, it was just a matter of running `npm install` for that error to disappear. It's safe to remove this stub now.

smartinellibenedetti commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/components/primeVue/PtAutoComplete/tests/PtAutoComplete.spec.ts:
```diff
@@ -0,0 +1,72 @@
+    return wrapper;
+};
+describe("PtAutoComplete", () => {
+    it("renders the component and passes props correctly", async () => {
+        const wrapper = await createWrapper();
+        expect(wrapper.props("modelValue")).toBe("initial value");
+        expect(wrapper.props("suggestions")).toEqual([
+            { label: "Option 1", value: 1 },
+            { label: "Option 2", value: 2 },
+        ]);
+        expect(wrapper.props("name")).toBe("test-autocomplete");
+        expect(wrapper.props("optionLabel")).toBe("label");
+    });
```
blocker: this test needs to be adjusted to expect that autocomplete is rendering things correctly based on the props received by checking the rendered HTML, as discussed before

smartinellibenedetti reviewed (changes requested):
Please ensure that the component's appearance matches what was defined in figma.

smartinellibenedetti commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/components/primeVue/PtAutoComplete/tests/PtAutoComplete.spec.ts:
```diff
@@ -0,0 +1,77 @@
+import PtAutoComplete from "../PtAutoComplete.vue";
+import AutoComplete from "primevue/autocomplete";
+
+config.global.mocks = {
+  $primevue: {
+    config: {}
+    }
+  };
```
question: did you undo the changes in the setupTest because it didn't work with the file being referenced only on setupFiles hook? 

If that was the case, please undo this change, and bring back the changes you had in setupTests to the file jest.global-mocks.js. It was beneficial to load primeVue as a global configuration for the team.

jayas006 replied:
> question: did you undo the changes in the setupTest because it didn't work with the file being referenced only on setupFiles hook?
> 
> If that was the case, please undo this change, and bring back the changes you had in setupTests to the file jest.global-mocks.js. It was beneficial to load primeVue as a global configuration for the team.

I didn't undo your changes. I first made my changes in the setup file. However, after reviewing it I thought I wasn't supposed to modify jest.global-mocks.js or jest.config.js, so I reverted my changes and left the files as it was before. So what's currently in the file is the original one. If you would like to review the changes I made before reverting them, I would happy to do them so that you can compare

smartinellibenedetti replied:
Let me clarify my comment as it seems that there's a misunderstanding. My changes weren't affected, I am talking about the changes in this commit: https://github.com/rundeck/rundeck/pull/9511/commits/e5b1ccb1a054d3f1cf34b700d1eece2ab0c79a98

These changes were beneficial for the team, as they would prevent the need for doing additional configuration in each test file.

What I'm recommending is to bring them back, but instead of having them living in the file setupTest.js file, to have them in the jest.global-mocks.js file.

smartinellibenedetti reviewed (changes requested):
The component still does not reflect the correct colors as defined in figma. Please provide a screenshot from storybook in this PR once you've done the fixes, making sure that both CSS files with the name tokens-rundeck match their content.

jayas006 commented:
> The component still does not reflect the correct colors as defined in figma. Please provide a screenshot from storybook in this PR once you've done the fixes, making sure that both CSS files with the name tokens-rundeck match their content.

Thanks, I will share the screenshot. I have asked this question before that why component is not reflecting the colors defined in Figma. But I will review it one more time and share the screenshot

smartinellibenedetti commented:
> Thanks, I will share the screenshot. I have asked this question before that why component is not reflecting the colors defined in Figma. But I will review it one more time and share the screenshot

If you're referring to the question asked in slack, I quote my own answer:

>  there are 2 files with the same name tokens-rundeck.css that are currently in different locations (this is temporary for now until we nail the final structure). One of them you've added new tokens in one of the previous tasks you've worked on, but need to check that the other file also has those tokens. Let me know if it doesn't make sense
> the second file was added as part of the PR you're working on right now, so it wasn't possible to have added the original tokens in both places at once

At the time, you had replied that this made sense. So as we discussed previously, please ensure that both files contain the same content and provide a screenshot once corrected, to facilitate the review.

jayas006 commented:
> > Thanks, I will share the screenshot. I have asked this question before that why component is not reflecting the colors defined in Figma. But I will review it one more time and share the screenshot
> 
> If you're referring to the question asked in slack, I quote my own answer:
> 
> > there are 2 files with the same name tokens-rundeck.css that are currently in different locations (this is temporary for now until we nail the final structure). One of them you've added new tokens in one of the previous tasks you've worked on, but need to check that the other file also has those tokens. Let me know if it doesn't make sense
> > the second file was added as part of the PR you're working on right now, so it wasn't possible to have added the original tokens in both places at once
> 
> At the time, you had replied that this made sense. So as we discussed previously, please ensure that both files contain the same content and provide a screenshot once corrected, to facilitate the review.

Thanks for the clarification, I will make sure to check both files and provide a screenshot once corrected

jayas006 commented:
tokens-rundeck.css
<img width="321" alt="tokens-rundeck css" src="https://github.com/user-attachments/assets/e1defbaa-7698-4ee3-93af-e192e27ee6e4" />

stylesheet/tokens-rundeck.css
<img width="348" alt="Screenshot 2025-02-12 at 6 22 57 PM" src="https://github.com/user-attachments/assets/97e8ad9c-3696-4589-88fa-9adbea8a8bb1" />

jayas006 commented:
@smartinellibenedetti , I have added the screenshot and also updated the test. Please let me know if anything else I need to do. Thank you!

jayas006 commented:
> @smartinellibenedetti ,I have added the screenshots. The only difference I noticed is the spelling of "gray" and "grey." I have changed it to "grey" as it is in Figma. Also, not all content from the stylesheet tokens-rundeck.css is in tokens-rundeck.css. I don't remember the reason for this, but other than that, all content matches. I have also updated the jest.global-mocks file. Please let me know if there is anything else I need to do. Thank you!

smartinellibenedetti commented:
> > The component still does not reflect the correct colors as defined in figma. Please provide a screenshot from storybook in this PR once you've done the fixes, making sure that both CSS files with the name tokens-rundeck match their content.
> 
> Thanks, I will share the screenshot. I have asked this question before that why component is not reflecting the colors defined in Figma. But I will review it one more time and share the screenshot

Please as discussed before, the screenshot was meant to be of the styled component in storybook, not from the code. Also, I'd like to suggest to review the screenshots of the CSS files already attached, to find discrepancies.

jayas006 commented:
@smartinellibenedetti , Thanks for your patience, Attaching the screenshots from storybook 
<img width="358" alt="Screenshot 2025-02-13 at 1 26 29 PM" src="https://github.com/user-attachments/assets/a0b38080-be6e-4e98-8ffe-95f0f714c3e0" />
<img width="277" alt="Screenshot 2025-02-13 at 1 25 33 PM" src="https://github.com/user-attachments/assets/5297a5a7-8536-4d5b-a53f-e14753d48317" />

smartinellibenedetti commented on rundeckapp/grails-app/assets/stylesheets/tokens-rundeck.css:
```diff
@@ -1,12 +1,14 @@
 :root {
+    --colors-white: #FFFFFF;
+    --colors-black: #18181B;
```
Please remove these 2 lines as they won't be needed (one of them will be redundant)

smartinellibenedetti commented on rundeckapp/grails-app/assets/stylesheets/tokens-rundeck.css:
```diff
@@ -1,12 +1,14 @@
     --colors-gray-100: #FAFAFA;
     --colors-gray-300: #E4E4E7;
-    --colors-gray-500: #A1A1AA;
-    --colors-gray-600: #71717A;
-    --colors-black:  #18181B;
+    --colors-grey-500: #A1A1AA;
+    --colors-grey-600: #71717A;
+    --colors-grey-900:  #18181B;
```
these should have been kept as the original names (gray and black). I know that the figma has "grey" instead of "gray", but we need to keep the names matching what was originally defined by continuum, which is "gray" (theme/tokens.css)

smartinellibenedetti commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/components/primeVue/PtAutoComplete/PtAutoComplete.vue:
```diff
@@ -36,15 +55,39 @@ export default defineComponent({
+    &:hover{
+      border-color: var(--colors-blue-500);
+    }
+
```
the border color of blue-500  should also be applied to `:enabled:focus`, along with `box-shadow: 0 0 0 0.2rem var(--colors-blue-100);`.

The box shadow wasn't described within that part of the figma, but during the review I noticed the green outline, that can be addressed within the same selector.

smartinellibenedetti commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/components/primeVue/PtAutoComplete/PtAutoComplete.vue:
```diff
@@ -23,11 +30,23 @@ export default defineComponent({
       type: Array,
       default: () => [],
     },
+    defaultValue: {
+      type: String,
+      default: "",
+    },
+    name: {
+      type: String,
+      default: "",
+    },
+    optionLabel: {
+      type: String,
+      default: "label",
+    },
```
They weren't in the ticket but noticed during the review that 2 more props are needed: `invalid` which is a boolean, and `placeholder` which is a string. Could you please add them?

jayas006 replied:
> They weren't in the ticket but noticed during the review that 2 more props are needed: `invalid` which is a boolean, and `placeholder` which is a string. Could you please add them?

 I am adding it, And I have added placeholder and invalid prop (for invalid it's a :class binding) inside Autocomplete component.

smartinellibenedetti commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/theme/tokens-rundeck.css:
```diff
@@ -0,0 +1,19 @@
+:root {
```
Please make sure that the changes done in the other CSS file are also reflected here.

smartinellibenedetti reviewed (changes requested):
The reason screenshots were asked was to allow both the author and the reviewer to validate that the component had the correct appearance. 

For example, upon trying to validate that selecting the input shows the right colors, this is the outcome:
![image](https://github.com/user-attachments/assets/5ddd6e17-b5fd-4613-876f-14d71e2c33c0)

In the same way, to see how the component looks when invalid, it was necessary to pass the prop with the same name. While the ticket missed this info and the placeholder props, validating the implemented solution would have caught that.

Lastly, this PR is missing the CSS rules for when the input is focused (it should show a blue border while the user types).  

Please add these changes: the CSS for the focused state and the props invalid and placeholder.

jayas006 commented:
> The reason screenshots were asked was to allow both the author and the reviewer to validate that the component had the correct appearance.
> 
> For example, upon trying to validate that selecting the input shows the right colors, this is the outcome: ![image](https://private-user-images.githubusercontent.com/139791797/414924128-5ddd6e17-b5fd-4613-876f-14d71e2c33c0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDAwMDY4NzgsIm5iZiI6MTc0MDAwNjU3OCwicGF0aCI6Ii8xMzk3OTE3OTcvNDE0OTI0MTI4LTVkZGQ2ZTE3LWI1ZmQtNDYxMy04NzZmLTE0ZDcxZTJjMzNjMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxOVQyMzA5MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNjg3NDQwZTM2NzNlZWI5ZWQ3MjJlMmNlOGEwZmQ4NTE2ZTMzMjE1NTJkMTBmNDg1MzBhYjNlOWMwNjM5NzQwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Z978108UZ585Iid4BqzBP9RH_BLFEjAnasK3UtrJbfE)
> 
> In the same way, to see how the component looks when invalid, it was necessary to pass the prop with the same name. While the ticket missed this info and the placeholder props, validating the implemented solution would have caught that.
> 
> Lastly, this PR is missing the CSS rules for when the input is focused (it should show a blue border while the user types).
> 
> Please add these changes: the CSS for the focused state and the props invalid and placeholder.

Thanks Sarah, It's so nice to know about it. Thank you for explaining

smartinellibenedetti commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/components/primeVue/PtAutoComplete/PtAutoComplete.vue:
```diff
@@ -2,13 +2,22 @@
+    :optionLabel="optionLabel"
+    :name="name"
+    :placeholder="placeholder"
+    :class="{ 'p-invalid': invalid }"
```
```suggestion
    :invalid="invalid"
```

The component already has this logic inside of it to add the "p-invalid" class when invalid is true, so all it's required is to pass the prop. If it didn't have this logic, the statement you had added would be completely correct.

jayas006 replied:
> The component already has this logic inside of it to add the "p-invalid" class when invalid is true, so all it's required is to pass the prop. If it didn't have this logic, the statement you had added would be completely correct.

For learning and understanding, So If invalid is true then simply passing the prop is enough, and no logic is needed. However, if invalid is false then the logic should be added accordingly?

smartinellibenedetti replied:
Not exactly. I'm saying that the component inside primevue already has the logic to add the class "p-invalid" when the prop invalid is true, so all it is required is to pass the prop to it. 

If we had a special case and let's say we wanted to add a completely different class, ie "rundeck-autocomplete" when invalid is true, then the logic you had added would be correct (adjusting the class name of course):
`:class="{ 'rundeck-autocomplete': invalid }"`.

Whenever in doubts if the component has the behaviour already, the primevue documentation is super handy: 
https://primevue.org/autocomplete

smartinellibenedetti reviewed (commented):
One tiny fix and it's good to go!

jayas006 commented:
> One tiny fix and it's good to go!

Done 👍
